### PR TITLE
Enable simulation to run with multiple ncdm species

### PR DESCRIPTION
**Pull request by adrianbayer**
Added relevant functionality to compute the distribution function for multiple species of neutrinos, and perform the split.

### CODE REVIEW

File: libfastpm/thermalvelocity.c [hunk 1]
```diff
@@ -73,29 +73,52 @@ void pix2vec (int pix, double *vec, int n_side)
     vec[2] = z;
 }
 
+/* make params struct for functions we will integrate with gsl */
+typedef struct kernel_params {double m[3]; int n;} kernel_params;
```
Comment (rainwoodman) [inline libfastpm/thermalvelocity.c]: Please break this into lines.

File: libfastpm/thermalvelocity.c [hunk 2]
```diff
@@ -73,29 +73,52 @@ void pix2vec (int pix, double *vec, int n_side)
...
-    return x / (exp(x) + 1);
+    /* FD kernel in the case of multiple particles
+       over the full volume (i.e. no *x^2) */
+    struct kernel_params * params = (struct kernel_params *)p;
+
+    double result = fermi_dirac_kernel_vol_1(x);
+    // only loop from i=1. i=0 term has been included in result already (r=1)
+    for(int i = 1; i < params->n; i ++) {
+        double r = params->m[i] / params->m[0];
+        if (r > 1) fastpm_raise(-1, "Please input the heaviest ncdm particle first.\n");
```
Comment (rainwoodman) [inline libfastpm/thermalvelocity.c]: If you can, report the error before getting this deep in the call chain -- such that it crashes before the run started. If you can't find an easy way for this, then it's ok as is.

File: libfastpm/thermalvelocity.c [hunk 3]
```diff
@@ -73,29 +73,52 @@ void pix2vec (int pix, double *vec, int n_side)
...
 
+/* Needed for calculatin the velocity dispersion */
+double fermi_dirac_dispersion(double x, void * p)
+{
+     return x*x*x*x * fermi_dirac_kernel_vol(x, p);
+}
 
 
-void divide_fd(double *vel_table, double *mass, int n_shells, int lvk)
+void divide_fd(double *vel_table, double *mass, kernel_params * params, int n_shells, int lvk)
```
Comment (rainwoodman) [inline libfastpm/thermalvelocity.c]: Long functions you can reformat it by keeping 'void' as a line, break lines of function arguments at meaning group boundaries.

File: libfastpm/thermalvelocity.c [hunk 4]
```diff
@@ -287,52 +314,41 @@ _fastpm_ncdm_init_fill(FastPMncdmInitData* nid)    ///call in create.  no need f
...
-
-                double velocity_conversion_factor = 50.3 * (1. + nid->z) * (1./m);
-                for(d = 0; d < 3; d ++){
-                    nid->vel[s][d] = vel_table[j]*vec_table[i*3+d]*velocity_conversion_factor;
-                }
-                s++;
+            double m0 = nid->m_ncdm[0];
+            /* define mass st sum over split gives 1*/
+            nid->mass[s] = masstab[j] / (12.*n_side*n_side);
+            double velocity_conversion_factor = 50.3 * (1. + nid->z) * (1./m0);
```
Comment (rainwoodman) [inline libfastpm/thermalvelocity.c]: Add a comment / source of this formula and 50.3 (arxiv link or wikipedia?)

File: tests/ncdm.lua [hunk 5]
```diff
@@ -8,19 +8,20 @@ boxsize = 1000
 -------- Time Sequence ----
 -- linspace: Uniform time steps in a
 -- logspace: Uniform time steps in loga
-time_step = linspace(0.05,1.,11)
+--time_step = linspace(0.05,1.,41)
+time_step = logspace(math.log10(0.05), 0, 11)
```
Comment (rainwoodman) [inline tests/ncdm.lua]: Not for the code, but logspace may lose a lot of halos at low-z.

**Comment by rainwoodman** [comment]
I did some formatting suggestions, but didn't read through the logic -- my impression was you have scrutinized this heavily on the email thread.

**Comment by adrianbayer** [comment]
Okay, thanks! I'll make the changes when Cori is back.

File: libfastpm/thermalvelocity.c (hunk 2 above)
Comment (adrianbayer) [reply libfastpm/thermalvelocity.c]: I put the error check in /src/lua-runtime-fastpm.lua instead

**Pull request merged**